### PR TITLE
Voice Announcement System (website part)

### PR DESCRIPTION
**Pull request by monster860**
This is the website part of my voice announcement system thingy.

Note for alex the catboy enthusiast: environment variable GAME_VOICE_ANNOUNCE_DIR must be set to like a directory that's shared between the game server and this.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -5,6 +5,9 @@ ENV UWSGI_INI /srv/www/yogsite/uwsgi.ini
 COPY . /srv/www/yogsite
 COPY nginx.conf /etc/nginx/sites-available/ 
 
+RUN apt-get install -y ffmpeg
```
Comment (AshCorr) [inline Dockerfile]: Seems like current available version on Debian is ffmpeg v4.1.6 which is currently vulnerable to a RCE vulnerability, see https://www.cvedetails.com/cve/CVE-2021-30123/

**Review by AshCorr — CHANGES_REQUESTED** [review]

File: yogsite/modules/voice_announce/routes.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	time_since_created = datetime.utcnow().replace(tzinfo=timezone.utc) - created_date
+	if time_since_created > timedelta(seconds = 15):
+		return Response("Voice announcement URL expired", status=404)
+
+	return render_template("voice_announce/voice_announce.html", enable_robot_voice = data["is_ai"])
+
+@blueprint.route("/voice_announce/<string:id>/upload", methods=["POST"])
+@flask_csrf_ext.exempt
+def voice_announce_upload(id):
+	id = secure_filename(id)
```
Comment (AshCorr) [inline yogsite/modules/voice_announce/routes.py:41]: Awesome

File: Dockerfile (hunk 1 above)
Comment (monster860) [reply Dockerfile]: How do I improve this situation

Comment (AshCorr) [reply Dockerfile]: There isnt really an ideal solution until latest version of ffmpeg is moved into the stable branch.

You can use the experimental branch, but who knows what issues that might have.

```
RUN echo "deb http://deb.debian.org/debian experimental main" >>  /etc/apt/sources.list
RUN apt-get update
RUN apt-get install -y -t experimental ffmpeg
```

Or see if theres some alternative / workaround... I couldn't see any listing for a workaround on the CVE though. Maybe you could disable compression?

Comment (AshCorr) [reply Dockerfile]: Seems like there are solutions for converting to ogg format which can run in browser.

https://blog.addpipe.com/using-webaudiorecorder-js-to-record-audio-on-your-website/
https://github.com/higuma/web-audio-recorder-js
https://github.com/kbumsik/opus-media-recorder

Comment (monster860) [reply Dockerfile]: I still need to use ffmpeg to verify the duration of the sound clip to make sure the user didn't upload like a long-ass sound clip or something

Comment (monster860) [reply Dockerfile]: Alright after doing some testing I was able to get the experimental version of ffmpeg to install properly

File: yogsite/modules/voice_announce/routes.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+@blueprint.route("/voice_announce/<string:serversqlname>/<string:id>")
+def page_voice_announce(serversqlname, id):
+	for s in cfg.get("servers").values():
+		if s["sqlname"] == serversqlname:
+			server = s
+	res = topic_query(server, "", args = {
+		"voice_announce_query": id,
+		"key": server["comms_key"]
+	})
+	print(res["exists"])
```
Comment (qwertyquerty) [inline yogsite/modules/voice_announce/routes.py]: dont forget to get rid of these

**Comment by JamieD1** [comment]
Ash is going to kill me

**Pull request merged**